### ISSUE TITLE
KOReader crashes after switching to another app and then returning to KOReader on Motorola Z4, Android 9

### ISSUE THREAD

**Issue by theshimonmor** [koreader/koreader#12879]
* KOReader version: 2024.11
* Device: Motorola Z4, Android 9

#### Open KOReader, use it, press either home or overview buttons on Android's 3-button navigation at bottom of screen, open another app then switch back to KOReader using overview button and KOReader opens for about 1/2 second and then crashes and I'm back at Android home screen. 

[crash.log](https://github.com/user-attachments/files/18054475/crash.log)

**Comment by theshimonmor** [comment]
This same issue happened with the two previous versions of KOReader.

**Comment by pazos** [comment]
That's quite unexpected on a Motorola :/.

Please try to reproduce on latest nightly (from github, not f-droid!, you can have both installed at the same time!).

If you're able to reproduce we would like "full adb logs" while you reproduce the crash. That would mean:

1. get adb working between your computer and your device
2. issue `adb logcat`
3. reproduce the issue using steps in OP.
4. copy the output and paste here.

**Comment by theshimonmor** [comment]
[adblogcat.txt](https://github.com/user-attachments/files/18067153/adblogcat.txt)

I installed the nightly (koreader-android-arm64-v2024.11-57-g0386e2619_2024-12-08.apk) and ran logcat which is attached.

**Comment by pazos** [comment]
Thank you very, very much!

In main menu -> settings -> device -> lights ->timeout:

disable custom timeouts by enabling either "using system defaults" or "keep the screen on"

**Comment by Frenzie** [comment]
Agreed.

**Comment by theshimonmor** [comment]
> Thank you very, very much!
> 
> In main menu -> settings -> device -> lights ->timeout:
> 
> disable custom timeouts by enabling either "using system defaults" or "keep the screen on"
> 
> Exit the program. Go to android settings, fix the screen timeout, which must be borked. Start the program.
> 
> Try to reproduce.

When I go to main menu > settings > device there is no option for "lights".  See screenshot.
![Screenshot_20241209-114637](https://github.com/user-attachments/assets/ac20f35f-425c-4a88-9207-c65307548c0b)

**Comment by pazos** [comment]
Yay, sorry. Updated instructions. Memory failed me :)

**Comment by pazos** [comment]
Thanks for the feedback.

Still something is fishy with custom timeouts. From the log you attached:

`12-09 11:27:54.482 25367 25367 V Timeout : onResume: updating system timeout: 5`

I'll investigate that.

As for the original issue: it's weird. This is the first reproducible crash that affects a regular android device.

It would be wonderful to try to reproduce on the emulator (android 9, simulated notch) or get some help from device owners that can reproduce the ticket.

We mostly need the `tombstone` that contains details of the crash. It's path is logged when the app crashes. Sadly it isn't possible to retrieve it directly unless you're root.

Non root users should follow a generic approach that will pull a bunch of unrelated data: https://developer.arm.com/documentation/108035/0100/MTE-bug-reports/Capture-a-bug-report-using-Developer-Options-on-your-Android-device?lang=en

It's up to you to attached the bug report entirely (and the tombstone path!) or just the tombstone from the bug report.

**Comment by theshimonmor** [comment]
My phone does have root privileges so I can access the tombstone reports.

I crashed KOReader and a new tombstone report was created.  I will post it but can you tell me if it captures any sensitive information I should remove first?  Thanks.

**Comment by theshimonmor** [comment]
I read through the tombstone file and didn't see any private info so here it is.

[tombstone_07.txt](https://github.com/user-attachments/files/18081756/tombstone_07.txt)

**Comment by theshimonmor** [comment]
Another thing I just noticed is that if I'm in the file browser portion of KOReader the crashes don't happen.  It only happens when I have a book open.  I can be in the file browser, switch to another app, and return to the KOReader file browser with no crash.  Crash only happens when a book is open.

**Comment by pazos** [comment]
> Crash only happens when a book is open.

Sounds a bit like https://github.com/koreader/koreader/issues/12659. Please try disabling partial rendering.

Anyhow the crash is totally different.

I'll try to reproduce on a emulator this weekend. Hopefully it is some regresion affecting how we handle the notch on android Pie. That should be easy to fix.

If your issue goes away disabling partial rendering let us know

**Comment by theshimonmor** [comment]
> Sounds a bit like #12659. Please try disabling partial rendering.
> 

Disabled partial rendering globally, exited app, opened app and book but it still crashes.  Tried on nightly version and 2024.11 f-droid versions.

**Comment by theshimonmor** [comment]
Did the tombstone file reveal anything?

**Comment by pazos** [comment]
Yup.

The crash, `signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)` is not an unmanaged exception raised from the JVM but a crash on libc and the backtrace doesn't point to any of our libraries.

Without the device that produces the crash and a debugger I cannot tell you more :/.

The crash happens when managed code has executed the `onResume` callback, just when it is creating and showing a surface.

That's why I wonder if it's something reproducible on plain Android Pie (with a notch -.- Does your device have a notch, right?. It says so in the logs.). 

Until now yours is the first report. Hopefully somebody else should be able to repro soon.

**Comment by theshimonmor** [comment]
Ok. Thanks for the explanation.

And by "notch"...I assume you mean the camera notch, correct? If so, yes this phone has a camera notch.

**Comment by hugleo** [comment]
> This same issue happened with the two previous versions of KOReader.

Before that things were working normally?

**Comment by theshimonmor** [comment]
> > This same issue happened with the two previous versions of KOReader.
> 
> Before that things were working normally?

It's never worked normally.  I only started using the app two versions ago.  Love the app so I lived with the crashing hoping an update might fix it.  Thought I would finally bring the issue up.

**Comment by hugleo** [comment]
Does it crash only when you open an epub book? Does it also crash for other file types like PDFs or even when no file is opened at all?

**Comment by poire-z** [comment]
A few things to test, if it's related to EPUB/ReaderRolling:

Sswitch from page mode to scroll mode in the bottom menu: there is a self:updateBatteryState() in ReaderRolling:onResume() which does stuff in page mode - but nothing in scroll mode. Does it change anything ?

After opening an EPUB, go back to file browser, so that the EPUB cache file is saved. Re-open this same EPUB a few times, and notice that it closes fast (no cache file rewritten as the previous one is valid) and re-opens fast (from a valid cache file, no re-rendering needed).
Then, try switching to another Android app, and back: still the crash ?

**Comment by poire-z** [comment]
Good, thanks.

>  Is there anything further that can be done as I'm not a fan of "continuous" mode?

In page mode, this updateBatteryState also does nothing when the top/alt status bar is disabled.
I expect you have it enabled?
Try disabling it:
![image](https://github.com/user-attachments/assets/dac6c220-66fa-41bc-9eb0-e5b9d3ad0d51)

Re-test.
If it does not crash, good, we again have found the culprit. We need to investigate further:

Re-enable it, and now go to the top menu ![image](https://github.com/user-attachments/assets/de5ea9df-8a70-444b-8312-26ee6b990ec5)
and try to disable everything, and re-test.
If it does not crash, try to re-enable item and features one by one, until you hopefully find the single thing responsible for the crashy behaviour.
Either you will live without that thing, or we may get an idea of why it can causes a crash and maybe an idea to avoid it.

**Comment by theshimonmor** [comment]
> In page mode, this updateBatteryState also does nothing when the top/alt status bar is disabled. I expect you have it enabled? Try disabling it:
> 
> Re-test. If it does not crash, good, we again have found the culprit. We need to investigate further:
> 

Alt Status Bar is enabled on my device.  I disabled it and there is no crash.

> Re-enable it, and now go to the top menu and try to disable everything, and re-test. If it does not crash, try to re-enable item and features one by one, until you hopefully find the single thing responsible for the crashy behaviour. Either you will live without that thing, or we may get an idea of why it can causes a crash and maybe an idea to avoid it.

I started disabling features under alt status bar and two features seem to cause crashes when enabled: "Auto refresh" and "Current time".  With both of those features off and alt status bar enabled the crashes stop.

![auto refresh](https://github.com/user-attachments/assets/28455bb2-e4d0-4712-bfe8-77f9e3b01ea2)

I can live without those features so I'm satisfied.  I'll keep testing to make sure the problem is actually solved and will report back if I get another crash.  But it appears as if you found the issue.

Thank you @poire-z for finding the solution and to @pazos and @hugleo for taking the time help.

**Comment by poire-z** [comment]
Thanks for the tests.
To be sure I understand you: it crashes too with Current time on - and Autorefresh off ?
And it crashes too with Current time off - and Autorefresh on ?

**Comment by theshimonmor** [comment]
> Thanks for the tests. To be sure I understand you: it crashes too with Current time on - and Autorefresh off ? And it crashes too with Current time off - and Autorefresh on ?

Correct.  If either or both options are enabled, the app will crash when switching apps.

**Comment by theshimonmor** [comment]
I've also found that having "battery status" enabled in the alt status bar will cause the occasional crash.  So three items, whether enabled singly or together, cause crashes: "battery status", "Auto refresh" and "Current time".

**Comment by theshimonmor** [comment]
Haven't had a crash since turning off "battery status", "Auto refresh" and "Current time" on the Alt Status Bar.  Been four days and about 10 hours of use.  Case closed.  Thanks again for the help.

**Comment by theshimonmor** [comment]
Ok. Want sure if the convention.

**Comment by poire-z** [comment]
It felt this is a device specific issue - to be confirmed by people using the top status bar with one or more of these 3 items enabled that they get no such issue.
If all is fine on most other Android devices - unless one gets some intuition about what may be at play, it probably won't be addressed. An android device life is short :)

**Comment by Frenzie** [comment]
It's fine on all of mine anyway. (Both older and newer.)
